### DISCUSSION TITLE
Forms support

### DISCUSSION THREAD

**Discussion by nerdoc** [tetra-framework/tetra#28] [Q&A] [answered]
Just a small question, I know that Django Forms support is still planned and not implemented yet. But even using Forms manually is a problem for me...

```python
class PersonForm(forms.ModelForm):
    class Meta:
        model = Person
        fields="__all__"

class MyComponent(Component):
    form: ModelForm|None = None

    def do_something(self):
        p = Person.objects.get(pk=xyz)
        self.form = PersonForm(data=p)
```
I can't get the form into the template, as (in the browser) Alpine complains:
```
Alpine Expression Error: Server responded with an error 500 (Internal Server Error)
Expression: "do_something()"
```
Django complains:
```
...
_pickle.PicklingError: Can't pickle <function paginator_number at 0x7f5e35106560>: it's not the same object as django.contrib.admin.templatetags.admin_list.paginator_number
```

I don't even have Pagination in the works...

What am I doing wrong? Can forms not be pickled - ok - but how do I get them to the browser then? Just over the plain old views? Can't I put them over the wire using a Tetra component?

EDIT: I can render a component, with a form as attribute, if I instantiate that form in the `load()` method. If I do it in e.g. another method that is called from Tetra's frontend side by e.g. a button, the pickling error is raised, with different kinds of objects, like:
```
_pickle.PicklingError: Can't pickle <function capfirst at 0x7fa152f0bc70>: it's not the same object as django.utils.text.capfirst
```

**Comment by nerdoc** [comment]
I have a (temporary?) solution ATM. I am using `self.form = PersonForm` for an empty form - this work well e.g. with `{% crispy form %} to display the form correctly.
I don't even know why, as the form *should* be instantiated - but it is never.

I'm lost.

**Comment by samwillis** [comment]
There is a known issues with Django Forms that prevents them from being pickle-able.

> class MyComponent(Component):
>     form: ModelForm|None = None
> 
>     def do_something(self):
>         p = Person.objects.get(pk=xyz)
>         self.form = PersonForm(data=p)
> ```

With the this code you attach a bound form to the component, sadly this doesn't yet work as when the method returns the state of the component is picked.

`load()` is a special case, it tracks all attributes that are modified and excludes them from the pickled state. The arguments passed to `load()` are picked, and the load method is re-run when the components state is unpicked.

> I have a (temporary?) solution ATM. I am using self.form = PersonForm for an empty form

I assume thats within the `load()`? It would have been excluded from the picked state.

--

I have a few ideas of how to add form support, the simplest is to exclude any Form instances from the picked state (this is possible using Pickle hooks Tetra uses in state.py), when unpicked they would be `None` 😬. You then rely on the user re-initiating a form in each method, I think this is probably a no go, too may ways a user can be tripped up.

Ideally I want to find a way to pickle forms, this should again be possible if we can track the arguments passed to the form to initiate it. We may require a Form subclass for this - I'm looking into it.

**Reply by nerdoc** [reply]
Ah, ok. I searched a bit in Unicorn's code again, and they do it, AFAICT by instantiating the form each time automatically, see [here](https://github.com/adamghill/django-unicorn/blob/edaa0c5c9bdc49280400eb6b606e9c66be4c5800/django_unicorn/components/unicorn_view.py#L484) . this may not be very effective in terms of speed, but convenient.

**Comment by nerdoc** [comment, marked as answer]
Forms are supported in 0.1.2 onwards.
